src/plugins/main.js

```diff
@@ -118,25 +120,44 @@ export default async (context) => {
     }
   }
 
-  const loadAndSetLocale = async locale => {
-    const { loadLanguageAsync } = require('./utils')
+  const loadAndSetLocale = async (newLocale, init = false) => {
+    // Abort if different domains option enabled
+    if (app.i18n.differentDomains) {
+      return
+    }
 
-    let messages
+    // Abort if newLocale did not change
+    if (newLocale === app.i18n.locale) {
+      return
+    }
 
+    const oldLocale = app.i18n.locale
+    app.i18n.beforeLanguageSwitch(oldLocale, newLocale)
+    if (useCookie && !init) {
+      app.i18n.setLocaleCookie(newLocale)
+    }
+    // Lazy-loading enabled
     if (lazy) {
-      // Load fallback locale.
-      if (app.i18n.fallbackLocale && locale !== app.i18n.fallbackLocale) {
-        await loadLanguageAsync(context, app.i18n.fallbackLocale)
-      }
-
-      messages = await loadLanguageAsync(context, locale)
+      const { loadLanguageAsync } = require('./utils')
+      const messages = await loadLanguageAsync(context, newLocale)
+      app.i18n.locale = newLocale
+      app.i18n.onLanguageSwitched(oldLocale, newLocale)
+      await syncVuex(newLocale, messages)
```

Comment:
Those three lines above can be shared, just need to load messages in `if` branches.